src/Raven.Server/Documents/AI/Embeddings/QueryEmbeddingsBatchingWorker.cs

```diff
@@ -0,0 +1,378 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.Embeddings;
+using Raven.Client.Documents.Operations.AI;
+using Raven.Server.Background;
+using Raven.Server.Config;
+using Raven.Server.Config.Categories;
+using Raven.Server.Documents.ETL.Providers.AI.Embeddings;
+using Sparrow.Server.Logging;
+
+namespace Raven.Server.Documents.AI.Embeddings
+{
+    public class QueryEmbeddingsBatchingWorker(string databaseName,
+        AiConfiguration configuration,
+#pragma warning disable SKEXP0001
+        (AiConnectionString ConnectionString, ITextEmbeddingGenerationService Instance) service,
+#pragma warning restore SKEXP0001
+        SemaphoreSlim concurrencyLimiter,
+        RavenLogger logger,
+        CancellationToken shutdown)
+        : BackgroundWorkBase(databaseName, logger, shutdown)
+    {
+        private readonly string _databaseName = databaseName;
+
+        private readonly ConcurrentQueue<QueryEmbeddingsBatchRequest> _requestQueue = new();
+        private readonly Stopwatch _batchTimer = new();
+
+        // Flag that indicates the service is being disposed
+        private volatile bool _workerShuttingDown;
+        private int _activeOperations;
+
+        public Task<ReadOnlyMemory<float>[]> EnqueueRequestAsync(IList<string> values, CancellationToken cancellationToken)
+        {
+            var request = new QueryEmbeddingsBatchRequest(values, callerToken: cancellationToken, workerToken: CancellationToken);
+
+            if (_workerShuttingDown)
+                return request.CancelWithShutdownMessage();
+
+            bool wasEmpty = _requestQueue.IsEmpty;
+            _requestQueue.Enqueue(request);
+
+            // Start the batch timer if this is the first request
+            if (wasEmpty)
```

Comment:
This is *really* racy. Entirely possibly that you have `IsEmpty` false, but between checking and `Enqueue`, you did everything, so this is now _not_ waking up the timer, and you hang, basically.